packages/next/next-server/server/api-utils.ts

```diff
@@ -245,6 +256,171 @@ export function sendJson(res: NextApiResponse, jsonBody: any): void {
+  }
+
+  const getCookies = getCookieParser(req)
+  let cookies: NextApiRequestCookies
+  try {
+    cookies = getCookies()
+  } catch {
+    // TODO: warn
+    return false
+  }
+
+  const hasBypass = COOKIE_NAME_PRERENDER_BYPASS in cookies
+  const hasData = COOKIE_NAME_PRERENDER_DATA in cookies
+
+  // Case: neither cookie is set.
+  if (!(hasBypass || hasData)) {
+    return false
+  }
+
+  // Case: one cookie is set, but not the other.
+  if (hasBypass !== hasData) {
+    clearPreviewData(res as NextApiResponse)
+    return false
+  }
+
+  // Case: preview session is for an old build.
+  if (cookies[COOKIE_NAME_PRERENDER_BYPASS] !== options.previewModeId) {
+    clearPreviewData(res as NextApiResponse)
+    return false
+  }
+
+  const tokenPreviewData = cookies[COOKIE_NAME_PRERENDER_DATA]
+
+  const jsonwebtoken = require('jsonwebtoken') as typeof import('jsonwebtoken')
+  let encryptedPreviewData: string
+  try {
+    encryptedPreviewData = jsonwebtoken.verify(
+      tokenPreviewData,
+      options.previewModeSigningKey
+    ) as string
+  } catch {
+    // TODO: warn
+    clearPreviewData(res as NextApiResponse)
+    return false
+  }
+
+  const decryptedPreviewData = decryptWithSecret(
+    Buffer.from(options.previewModeEncryptionKey),
+    encryptedPreviewData
+  )
+
+  try {
+    // TODO: strict runtime type checking
+    const data = JSON.parse(decryptedPreviewData)
+    // Cache lookup
+    Object.defineProperty(req, SYMBOL_PREVIEW_DATA, {
+      value: data,
+      enumerable: false,
+    })
+    return data
+  } catch {
+    return false
+  }
+}
+
+function setPreviewData<T>(
+  res: NextApiResponse<T>,
+  data: object | string, // TODO: strict runtime type checking
+  options: {
+    maxAge?: number
+  } & __ApiPreviewProps
+): NextApiResponse<T> {
+  if (
+    typeof options.previewModeId !== 'string' ||
+    options.previewModeId.length < 16
+  ) {
+    throw new Error('invariant: invalid previewModeId')
+  }
+  if (
+    typeof options.previewModeEncryptionKey !== 'string' ||
+    options.previewModeEncryptionKey.length < 16
+  ) {
+    throw new Error('invariant: invalid previewModeEncryptionKey')
+  }
+  if (
+    typeof options.previewModeSigningKey !== 'string' ||
+    options.previewModeSigningKey.length < 16
+  ) {
+    throw new Error('invariant: invalid previewModeSigningKey')
+  }
+
+  const jsonwebtoken = require('jsonwebtoken') as typeof import('jsonwebtoken')
+
+  const payload = jsonwebtoken.sign(
+    encryptWithSecret(
+      Buffer.from(options.previewModeEncryptionKey),
+      JSON.stringify(data)
+    ),
+    options.previewModeSigningKey,
+    {
+      algorithm: 'HS256',
+      ...(options.maxAge !== undefined
+        ? { expiresIn: options.maxAge }
+        : undefined),
+    }
+  )
+
+  const { serialize } = require('cookie') as typeof import('cookie')
+  const previous = res.getHeader('Set-Cookie')
+  res.setHeader(`Set-Cookie`, [
+    ...(typeof previous === 'string' ? [previous] : []),
```

Comment:
```suggestion
    ...(typeof previous === 'string' ? [previous] : []),
    ...(Array.isArray(previous) ? previous : [])
```
`previous` could also be an array right?

Reply:
fixed